pkg/controller/controller_test.go

```diff
@@ -4803,11 +5448,28 @@ func TestProvisionFromPVC(t *testing.T) {
 			if tc.cloneUnsupported {
 				pluginCaps, controllerCaps = provisionCapabilities()
 			}
+			var volumeSource csi.VolumeContentSource_Volume
 			if !tc.expectErr {
-				volumeSource := csi.VolumeContentSource_Volume{
-					Volume: &csi.VolumeContentSource_VolumeSource{
-						VolumeId: tc.volOpts.PVC.Spec.DataSource.Name,
-					},
+				if tc.xnsEnabled {
+					if tc.volOpts.PVC.Spec.DataSourceRef != nil {
+						volumeSource = csi.VolumeContentSource_Volume{
+							Volume: &csi.VolumeContentSource_VolumeSource{
+								VolumeId: tc.volOpts.PVC.Spec.DataSourceRef.Name,
+							},
+						}
+					} else if tc.volOpts.PVC.Spec.DataSource != nil {
+						volumeSource = csi.VolumeContentSource_Volume{
+							Volume: &csi.VolumeContentSource_VolumeSource{
+								VolumeId: tc.volOpts.PVC.Spec.DataSource.Name,
+							},
+						}
+					}
+				} else {
```

Comment:
Can this be collapsed into 2 conditions?

Reply:
Fixed.